File: routes/reviews.ts
// Review Routes
import { Express } from "express";
import { Server } from "socket.io";

export function registerReviewRoutes(app: Express, db: any, io: Server) {
  app.get("/api/reviews", (req, res) => {
    try {
      res.json(db.prepare("SELECT * FROM reviews ORDER BY id DESC").all());
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });

  app.get("/api/reviews/my/:userId", (req, res) => {
    try {
      res.json(db.prepare("SELECT * FROM reviews WHERE user_id = ?").all(parseInt(req.params.userId)));
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });

  app.get("/api/reviews/summary", (req, res) => {
    try {
      const reviews = db.prepare("SELECT * FROM reviews ORDER BY id DESC").all();
      const total = reviews.length;
      const avgRating = total > 0 ? reviews.reduce((s: number, r: any) => s + r.overall_rating, 0) / total : 0;
      const wasteNone = reviews.filter((r: any) => r.waste_level === 'none').length;
      const wasteLot = reviews.filter((r: any) => r.waste_level === 'lot').length;
      res.json({ success: true, data: { total, avgRating: Math.round(avgRating * 10) / 10, wasteNone, wasteLot } });
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });

  app.post("/api/reviews", (req, res) => {
    try {
      const { user_id, meal_type, overall_rating, comment, waste_level, items, student_name } = req.body;
      const today = new Date().toISOString().split('T')[0];
      const info = db.prepare("INSERT INTO reviews (user_id, date, meal_type, overall_rating, comment, waste_level, student_name) VALUES (?, ?, ?, ?, ?, ?, ?)").run(user_id, today, meal_type, overall_rating, comment, waste_level, student_name || 'Anonymous');
      const reviewId = info.lastInsertRowid;
      if (items && Array.isArray(items)) {
        items.forEach((item: any) => {
          db.prepare("INSERT INTO item_reviews (review_id, food_item_id, taste, quantity, quality) VALUES (?, ?, ?, ?, ?)").run(reviewId, item.id, item.taste, item.quantity, item.quality);
        });
      }
      io.emit("new_review", { user_id, meal_type, overall_rating, student_name });
      if (overall_rating <= 2) {
        io.emit("low_rating_alert", { student_name, meal_type, rating: overall_rating, comment });
      }
      res.json({ success: true });
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });
}

File: backend/routes/reviews.py
from fastapi import APIRouter
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
from database import db

router = APIRouter(prefix="/api", tags=["reviews"])

# Socket.IO instance will be set by main.py
sio = None


class ItemReview(BaseModel):
    id: int
    taste: Optional[int] = 3
    quantity: Optional[int] = 3
    quality: Optional[int] = 3


class ReviewCreate(BaseModel):
    user_id: int
    meal_type: str
    overall_rating: int
    comment: Optional[str] = ""
    waste_level: Optional[str] = "none"
    items: Optional[List[ItemReview]] = None
    student_name: Optional[str] = "Anonymous"


@router.get("/reviews")
def list_reviews():
    reviews = db.get_all("reviews")
    return sorted(reviews, key=lambda r: r.get("id", 0), reverse=True)


@router.get("/reviews/my/{user_id}")
def my_reviews(user_id: int):
    return db.filter("reviews", user_id=user_id)


@router.get("/reviews/summary")
def review_summary():
    reviews = db.get_all("reviews")
    total = len(reviews)
    avg = round(sum(r.get("overall_rating", 0) for r in reviews) / total, 1) if total > 0 else 0
    waste_none = sum(1 for r in reviews if r.get("waste_level") == "none")
    waste_lot = sum(1 for r in reviews if r.get("waste_level") == "lot")
    return {"success": True, "data": {"total": total, "avgRating": avg, "wasteNone": waste_none, "wasteLot": waste_lot}}


@router.post("/reviews")
async def create_review(review: ReviewCreate):
    today = datetime.now().strftime("%Y-%m-%d")
    rid = db.insert("reviews", {
        "user_id": review.user_id, "date": today, "meal_type": review.meal_type,
        "overall_rating": review.overall_rating, "comment": review.comment,
        "waste_level": review.waste_level, "student_name": review.student_name or "Anonymous",
        "created_at": datetime.now().isoformat(),
    })
    if review.items:
        for item in review.items:
            db.insert("item_reviews", {
                "review_id": rid, "food_item_id": item.id,
                "taste": item.taste, "quantity": item.quantity, "quality": item.quality,
            })
    # emit socket events if sio is available
    if sio:
        await sio.emit("new_review", {
            "user_id": review.user_id, "meal_type": review.meal_type,
            "overall_rating": review.overall_rating, "student_name": review.student_name,
        })
        if review.overall_rating <= 2:
            await sio.emit("low_rating_alert", {
                "student_name": review.student_name, "meal_type": review.meal_type,
                "rating": review.overall_rating, "comment": review.comment,
            })
    return {"success": True}
